README.md

```diff
@@ -4,3 +4,207 @@
 
 `ppx_factory` is an OCaml preprocessor to derive factory methods and default values from type
 definitions.
+
+## Overview
+
+Factories are functions that let you build test data while only specifying the parts that are
+relevant to your tests.
+
+`ppx_factory` alows you to derive such functions from type definitions.
+
+Let's take a very basic example. Consider the following type and function:
+```ocaml
+type person =
+  { first_name : string
+  ; middle_name : string option
+  ; last_name : string
+  ; age : int
+  ; hobbies : Hobby.t list
+  }
+
+let full_name {first_name; last_name; middle_name; _} =
+  match middle_name with
+  | None -> Printf.sprintf "%s %s" first_name last_name
+  | Some m_name -> Printf.sprintf "%s \"%s\" %s" first_name m_name last_name
+```
+
+When writing tests for the `full_name` function, you don't want to bother defining `age` and
+`hobbies` every time so you use a factory method that looks like:
+```ocaml
+val factory_person :
+  ?first_name: string ->
+  ?middle_name: string ->
+  ?last_name: string ->
+  ?age: int ->
+  ?hobbies: Hobby.t list ->
+  unit ->
+  person
+```
+
+and let's you write your test in a more concise manner:
+```ocaml
+test
+  ~input:(factory_person ~first_name:"John" ?middle_name:None ~last_name:"Doe" ())
+  ~expected:"John Doe";
+test
+  ~input:(factory_person ~first_name:"Robyn" ~middle_name:"Rihanna" ~last_name:"Fenty" ())
+  ~expected:"Robyn \"Rihanna\" Fenty"
+```
+
+By adding the `[@@deriving factory]` attribute to your type definitions, `ppx_factory` will generate
+a single factory function for record types or one per constructor for variant types.
+
+See the [Detailed usage](https://github.com/cryptosense/ppx_factory#detailed-usage) section for
+further specification.
+
+## Installation and usage
+
+You can install `ppx_factory` using [opam](https://opam.ocaml.org):
+```
+$ opam install ppx_factory
+```
+
+If you're building your library or app with dune, add the following field to your `library`,
+`executable` or `test` stanza:
+```
+(preprocess (pps ppx_factory))
+```
+or simply add `ppx_factory` to your `preprocess` field if it's already there.
+
+You can now add the `factory` and/or `default` plugins to `[@@deriving ...]` attributes on type
+definitions to derive the corresponding values.
+
+## Detailed usage
+
+### factory
+
+You can use `[@@deriving factory]` to derive factory functions from type definitions both in module
+and module signatures eg both in your `.ml` and `.mli` files given that it's an explicit record or
+variant type definition.
+
+If you try to derive factory functions from an abstract type, like in the following example:
```

Comment:
I'm not sure this is really detailed usage - maybe a `known issues` section at the bottom?

Reply:
Or a `compatibility with ppx import` section?

Reply:
Yeah that'd probably be better, I'll move it!